Update keyless entity types topic (#1736)
aspnet/EntityFramework.Docs: entity-framework/core/modeling/query-types.md

---

> title: Keyless Entity Types - EF Core

Reviewer: should rename file.

Author: Don't want to deal with changing links/adding a redirect.
@divega thoughts?

Reviewer 2: It makes sense to rename the file, but I think it is a lower priority. If we do rename the file, we need to add a redirection on https://github.com/aspnet/EntityFramework.Docs/blob/master/.openpublishing.redirection.json

author: AndriySvyryd
ms.author: ansvyryd
ms.date: 02/26/2018
ms.assetid: 9F4450C5-1A3F-4BB6-AC19-9FAC64292AAD
uid: core/modeling/query-types
---
# Keyless Entity Types
> [!NOTE]
> This feature is new in EF Core 2.1. Prior to 3.0 they were known as Query Types

> In addition to normal entity types, an EF Core model can contain _keyless entity types_, which can be used to carry out database queries against data that doesn't contain key values.

Reviewer 3: Consider normal -> regular

> ## Keyless Entity Types Characteristics

Reviewer 3: I think our standard is for lower-case after the first word in non-toplevel headings

> Keyless entity types support many of the same mapping capabilities, like inheritance mapping and navigation properties. On relational stores, they can configure the target database objects and columns via fluent API methods or data annotations.

Reviewer: same capabilities as regular entity types?

However, they are different from normal entity types in that they:

- Cannot have a key defined.
- Are never tracked for changes in the _DbContext_ and therefore are never inserted, updated or deleted on the database.
- Are never discovered by convention.
- Only support a subset of navigation mapping capabilities, specifically:
  - They may never act as the principal end of a relationship.
  - They can only contain reference navigation properties pointing to entities.

>   - Entities cannot contain navigation properties to keyless entity types.

Author: Keyless entity types can't contain navigation properties to keyless entity types either.

- Need to be configured with `.HasNoKey()` method call.

> - May be mapped to a _defining query_. A defining query is a query declared in the model that acts as a data source for a keyless entity type.

Reviewer 2: Added a comment on https://github.com/aspnet/EntityFramework.Docs/issues/1223 to add links here once we have defining query documentation.

## Usage scenarios

Some of the main usage scenarios for keyless entity types are:

> - Serving as the return type for `FromSql()` queries.

Reviewer 2: There is no FromSql method anymore :smile:. Rather than adding the two methods:
```suggestion
- Serving as the return type for [raw SQL queries](xref:core/querying/raw-sql).
```

- Mapping to database views that do not contain a primary key.
- Mapping to tables that do not have a primary key defined.

> - Mapping to queries defined in the model.

Reviewer: last one is unclear. Probably should just omit.

Reviewer 2: Once we have a page on defining queries we can add a link here:
```suggestion
- Mapping to defining queries.
```

Author: I don't think this is clearer, we can add it when there's a page on it

## Mapping to database objects

> Mapping a keyless entity type to a database object is achieved using the `ToTable` or `ToView` fluent API. From the perspective of EF Core, the database object specified in this method is a _view_, meaning that it is treated as a read-only query source and cannot be the target of update, insert or delete operations. However, this does not mean that the database object is actually required to be a database view. It can alternatively be a database table that will be treated as read-only. Conversely, for normal entity types, EF Core assumes that a database object specified in the `ToTable` method can be treated as a _table_, meaning that it can be used as a query source but also targeted by update, delete and insert operations. In fact, you can specify the name of a database view in `ToTable` and everything should work fine as long as the view is configured to be updatable on the database.

Reviewer: Should also specify differences in migrations if any.

Reviewer 2: Also, this whole paragraph applies to regular entity types as well, so should it be replicated somewhere else in the docs? Or maybe moved into its own central place (in the mapping to tables section that @roji is working on in https://github.com/aspnet/EntityFramework.Docs/pull/1717) and then linked from here?

Reviewer 3: I'd say let's merge this as-is for now (we're mainly adjusting query types ->  keyless in this PR) and revisit refactoring this later.

## Example

> The following example shows how to use keyless entity type to query a database view.

Reviewer 3: ```suggestion
The following example shows how to use a keyless entity type to query a database view.
```

[!code-csharp[Main](../../../samples/core/QueryTypes/Program.cs#QueryType)]

Next, we configure the keyless entity type in _OnModelCreating_ using the `HasNoKey` API.

> We use standard fluent configuration API to configure the mapping for the keyless entity type:

Reviewer 3: ```suggestion
We use the fluent configuration API to configure the mapping for the keyless entity type:
```

Not sure why standard is needed here, but in any case `the` or `a` is missing.

> [!code-csharp[Main](../../../samples/core/QueryTypes/Program.cs#Configuration)]

Reviewer: Should update this code to be not so fluent.

Author: Could you elaborate? 😄

Reviewer: more like
```c#
modelBuilder.Entity<(b => 
{
    b.ToView("");
    b.Property().HasColumnName();
});
```

Reviewer 2: I think it might also be good to update it not to use deprecated API :trollface:
